connectivity/check/logger.go

```diff
@@ -0,0 +1,187 @@
+}
+
+// Start starts ConcurrentLogger internals in separate goroutines:
+// - collector: collects incoming test messages.
+// - printer: sends messages to the writer in corresponding order.
+func (c *ConcurrentLogger) Start(ctx context.Context) {
+	c.collectorStarted.Store(true)
+	go c.collector(ctx)
+	go c.printer()
+}
+
+// Stop closes incoming message channel and waits while all messages are printed.
+func (c *ConcurrentLogger) Stop() {
+	close(c.messageCh)
+	<-c.printerDoneCh
+}
+
+type message struct {
+	namespace string
+	testName  string
+	data      string
+	finish    bool
+}
+
+func (m message) nsTest() string {
+	return fmt.Sprintf("%s:%s", m.namespace, m.testName)
+}
+
+// Printf schedules message for the test to be printed.
+func (c *ConcurrentLogger) Printf(test *Test, format string, args ...interface{}) {
+	buf := &bytes.Buffer{}
+	if test.ctx.timestamp() {
+		mustFprintf(buf, timestamp())
+	}
+	mustFprintf(buf, format, args...)
+	c.messageCh <- message{
+		namespace: test.ctx.params.TestNamespace,
+		testName:  test.name,
+		data:      buf.String(),
+	}
+}
+
+// FinishTest schedules the final message for the test to be printed.
+// The message will be populated with the test log buffer if the test failed.
+func (c *ConcurrentLogger) FinishTest(test *Test) {
+	buf := &bytes.Buffer{}
+	if test.failed && test.logBuf != nil {
+		if _, err := io.Copy(buf, test.logBuf); err != nil {
+			panic(fmt.Errorf("failed to read from test log buffer: %w", err))
+		}
+	}
+	c.messageCh <- message{
+		namespace: test.Context().Params().TestNamespace,
+		testName:  test.Name(),
+		data:      buf.String(),
+		finish:    true,
+	}
+}
+
+func (c *ConcurrentLogger) collector(ctx context.Context) {
+	defer func() { c.collectorStarted.Store(false) }()
+	for {
+		select {
+		case m, open := <-c.messageCh:
+			if !open {
+				return
+			}
+			nsTest := m.nsTest()
+			c.nsTestMsgsLock.Lock()
+			nsTestMsgs, ok := c.nsTestMsgs[nsTest]
+			if !ok {
+				nsTestMsgs = make([]message, 0)
+				// use a separate goroutine to avoid deadlock if the channel
+				// buffer is full, printer goroutine will pull it eventually
+				go func() { c.nsTestsCh <- nsTest }()
+			}
+			c.nsTestMsgs[nsTest] = append(nsTestMsgs, m)
+			c.notifyPrinterIfWaiting(nsTest)
+			c.nsTestMsgsLock.Unlock()
+		case <-ctx.Done():
+			close(c.messageCh)
+			return
+		}
+	}
+}
+
+func (c *ConcurrentLogger) notifyPrinterIfWaiting(nsTest string) {
+	if c.printerWaitValue == nsTest {
+		c.printerWaitValue = ""
+		c.printerWaitCh <- struct{}{}
+	}
+}
+
+func (c *ConcurrentLogger) printer() {
+	// read messages while the collector is working
+	for c.collectorStarted.Load() {
+		// double-check if there are new messages to avoid
+		// deadlock reading from the `nsTestsCh` channel
+		if c.collectedTestCount() > c.nsTestCount {
+			c.printTestMessages(<-c.nsTestsCh)
+		}
+	}
+	// collector stopped but there still might be messages to print
+	for c.collectedTestCount() > c.nsTestCount {
```

Comment:
I think this condition (and the same one above), would be more readable if reversed. What do you think?

`for c.nsTestFinishedCount < c.collectedTestCount()`

Reply:
Makes sense! Done, thanks!